cache-openai-langchain.mdx

```diff
@@ -0,0 +1,517 @@
+// File: index.js
+
+// ...
+
+// XataCache class built on top of LangChain's BaseCache
+const { XataCache } = require('./langchain-xata-cache.js');
+
+const { ChatOpenAI } = require('@langchain/openai');
+
+const { getXataClient } = require('./xata');
+
+// Global Xata Client Instance
+const xata = getXataClient();
+
+// ...
+
+app.post('/query-cache-method', async (req, res) => {
+  const { input } = req.body;
+
+  // Use LangChain Xata Cache Adapter
+  const model = new ChatOpenAI({
+    cache: new XataCache({ client: xata })
+  });
+
+  console.time();
+
+  // Perform a cached-response first AI model invocation
+  const response = await model.invoke(input);
+
+  console.timeEnd();
+
+  res.write(response.content);
+  res.end();
+});
+
+// ...
+```
+
+The code above, starts with importing the `XataCache` class that we created earlier. Then, it creates an endpoint `/query-cache-method` which can be `POST`-ed to. Then, it destructures `input` from the request body. Then, it creates a LLM via a `ChatOpenAI` instance with LangChain. The `cache` property of the ChatOpenAI model is set to a new instance of `XataCache` using the existing global Xata client. Then, it invokes the model to respond to the user query. Finally, the content of the response by OpenAI is returned to the user.
+
+Great! Now, let’s test if responses are being cached by executing the following command in your terminal:
+
+```bash
+for i in {1..9}; do
+    curl --location 'http://localhost:3005/query-cache-method' \
+    --header 'Content-Type: application/json' \
+    --data '{
+        "input": "Who are you?"
+    }'
+    echo "Execution $i completed."
+done
+```
+
+The output in your terminal window would indicate the time (approximately) it took to respond to the user queries. An immediate decrease in the response times after the very first one indicates cached responses being returned for the user queries.
+
+![Multiple Queries Time Performance](/images/cache-openai-langchain/Screenshot_2024-03-08_at_3.12.34_AM.png)
+
+Awesome! You’re now successfully able to cache responses generated by OpenAI with Xata using the LangChain Caching Layer.
+
+## Final Thoughts
+
+In this guide, you learned different ways of caching OpenAI API responses using LangChain callbacks and the caching layer. Using Xata, you’re able to interact with a Postgres serverless database to build a set of cache OpenAI API responses, and utilize Xata’s search capabilities to respond to user queries faster using those cache responses.
+
+For more detailed insights, explore the references cited in this post.
+
+| Resource                        | Link                                                          |
+| ------------------------------- | ------------------------------------------------------------- |
+| GitHub Repo                     | https://github.com/rishi-raj-jain/xata-cache-openai-langchain |
+| Caching - LangChain Chat Models | https://js.langchain.com/docs/modules/model\_io/chat/caching  |
```

Comment:
can we update the link to https://js.langchain.com/docs/modules/model_io/chat/caching instead?